src/Computation/BitsExtensions.cs

```diff
@@ -0,0 +1,106 @@
+﻿using System;
+using System.Numerics;
+using Microsoft.Z3;
+using Symbolica.Expression;
+
+namespace Symbolica.Computation
+{
+    internal static class BitsExtensions
+    {
+        public static FPSort GetSort(this Bits self, Context context)
+        {
+            return (uint) self switch
+            {
+                16U => context.MkFPSort16(),
+                32U => context.MkFPSort32(),
+                64U => context.MkFPSort64(),
+                80U => context.MkFPSort(15U, 65U),
+                128U => context.MkFPSort128(),
+                _ => throw new Exception($"Floating-point size {self} is unsupported.")
+            };
+        }
+
+        public static BigInteger GetNan(this Bits self, Context context)
+        {
+            var sort = self.GetSort(context);
+            var nan = ((BigInteger.One << ((int) sort.EBits + 2)) - BigInteger.One) << ((int) sort.SBits - 2);
+
+            return ConstantUnsigned.Create(self, nan).Integer;
+        }
+
+        public static FPExpr ParseNormalFloat(this Bits self, Context context, string value)
+        {
+            var sort = self.GetSort(context);
+            var (sign, significand, exponent) = ParseDecimal(value);
+
+            return significand.IsZero
+                ? context.MkFPZero(sort, sign)
+                : CreateFloat(context, sort, sign, significand, exponent);
+        }
+
+        private static FPExpr CreateFloat(Context context, FPSort sort, bool sign, BigInteger significand, int exponent)
+        {
+            var (numerator, denominator) = exponent < 0
+                ? (significand, BigInteger.Pow(10, -exponent))
+                : (significand * BigInteger.Pow(10, exponent), BigInteger.One);
+
+            var precision = (int) sort.SBits;
+            var scale = precision - 1L;
+
+            var lower = BigInteger.One << (precision - 1);
+            while (numerator / denominator < lower)
+            {
+                numerator <<= 1;
+                --scale;
+            }
+
+            var upper = BigInteger.One << precision;
+            while (upper <= numerator / denominator)
+            {
+                denominator <<= 1;
+                ++scale;
+            }
+
+            var quotient = numerator / denominator;
+            var remainder = numerator % denominator;
+            var midpoint = denominator >> 1;
+
+            var (rounded, shift) = remainder > midpoint || remainder == midpoint && !quotient.IsEven
+                ? quotient == upper - BigInteger.One
+                    ? (lower, scale + 1L)
+                    : (quotient + BigInteger.One, scale)
+                : (quotient, scale);
+
+            return context.MkFP(sign, shift, (ulong) (rounded % lower), sort);
+        }
```

Comment:
I really, really wanted to avoid having to write anything like this, but unfortunately I don't see any way around it. To be honest, the implementation is not actually that difficult, thanks to the arbitrary precision magic of `BigInteger`, and I have written loads of tests which should cover all the branches and edge cases, but still annoying that I had to do this at all. If only I had something that could execute it symbolically for various precisions ...

Reply:
For future reference, [source](https://www.exploringbinary.com/correct-decimal-to-floating-point-using-big-integers/).